html/inc/consent.inc

```diff
@@ -0,0 +1,70 @@
+<?php
+// This file is part of BOINC.
+// http://boinc.berkeley.edu
+// Copyright (C) 2017 University of California
+//
+// BOINC is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// BOINC is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+
+// functions dealing with the consent and consent_type tables.
+
+include_once("../inc/boinc_db.inc");
+include_once("../inc/util.inc");
+
+function consent_to_a_policy($user, $consent_name, $consent_flag, $consent_not_required, $source, $ctime = 0) {
+    $myn = BoincDb::escape_string($consent_name);
+    $mys = BoincDb::escape_string($source);
+    if ($ctime==0) {
+        $mytime = $user->create_time;
+    }
+    else {
+        $mytime = $ctime;
+    }
+    return BoincConsent::insert(
+        "(id, userid, consent_name, consent_time, consent_flag, consent_not_required, source) " .
+        "values(0, $user->id, '$myn', $mytime, $consent_flag, $consent_not_required, '$mys')"
+    );
+
+}
+
+function consent_after_login($user, $perm=true, $next_url = "") {
+    session_start();
+    $_SESSION['user'] = $user;
+    $_SESSION['perm'] = $perm;
```

Comment:
This code is the first time that BOINC uses session.  Up to this point, the website has been essentially stateless (i.e. all information needed to support the request is contained on the request or loaded from the database based on information in the request).  I have two questions about this.

1) Do any BOINC projects do load balancing using more than one web server to serve their site?  If not, then question 2 isn't important.
2) Are we expecting BOINC projects that have more than one web server to use some form of server affinity?

If there is a project using load balancing between servers, then caching the data in the session could become problematic if the users request winds up on a server that doesn't have this data cached.  

This is a pretty big architectural decision to go from stateless to stateful requests.  Perhaps it would be better to store the data you are putting in the session in a temp cookie so that we don't make this change now.

Reply:
I admit I did not think of the high-level implications of using a stateful architecture when writing this.

I (obviously) need some kind of temporary space to store the user information  while traversing multiple pages, so the SESSION seemed like an obvious candidate. I have used it on occasion for Drupal-Web development.

But if the same functionality can be made with a temporary cookie, then I'll refactor this section of the code.

Reply:
Yea - I would go the temp cookie route and just put the authenticator and perm flag into the temp cookie so that you have it available after the consent.

Reply:
I agree that if session was already and established pattern on the site then this would make sense, but since it is new there might be deployment cases that this where this could be a big surprise and result in some bug reports while they tried to figure out what is going on.

Reply:
FWIW, I concur.

Reply:
This has been fixed.